stdlib/LinearAlgebra/src/triangular.jl

```diff
@@ -1189,289 +1307,356 @@ end
+            bi -= A.data[i,j] * c[j]
         end
+        c[i] = aii \ bi
     end
-    return b
+    return c
 end
-function ldiv!(A::UnitUpperTriangular, b::AbstractVector)
-    require_one_based_indexing(A, b)
+function ldiv!(c::AbstractVector, A::UnitUpperTriangular, b::AbstractVector)
+    require_one_based_indexing(c, A, b)
     n = size(A, 2)
     if !(n == length(b))
         throw(DimensionMismatch("second dimension of left hand side A, $n, and length of right hand side b, $(length(b)), must be equal"))
     end
-    @inbounds for j in n:-1:1
-        bj = b[j]
-        for i in j-1:-1:1 # counterintuitively 1:j-1 performs slightly better
-            b[i] -= A.data[i,j] * bj
+    if !(n == length(c))
+        throw(DimensionMismatch("length of output c, $(length(c)), does not match length of right hand side b, $(length(b))"))
+    end
+    @inbounds for i in n:-1:1
+        bi = b[i]
+        for j in i+1:n
+            bi -= A.data[i,j] * c[j]
         end
+        c[i] = A[i,i] \ bi
     end
-    return b
+    return c
 end
-function ldiv!(A::LowerTriangular, b::AbstractVector)
-    require_one_based_indexing(A, b)
+function ldiv!(c::AbstractVector, A::LowerTriangular, b::AbstractVector)
+    require_one_based_indexing(c, A, b)
     n = size(A, 2)
-    if !(n == length(b))
+    if n != length(b)
         throw(DimensionMismatch("second dimension of left hand side A, $n, and length of right hand side b, $(length(b)), must be equal"))
     end
-    @inbounds for j in 1:n
-        iszero(A.data[j,j]) && throw(SingularException(j))
-        bj = b[j] = A.data[j,j] \ b[j]
-        for i in j+1:n
-            b[i] -= A.data[i,j] * bj
+    if n != length(c)
+        throw(DimensionMismatch("length of output c, $(length(c)), does not match length of right hand side b, $(length(b))"))
+    end
+    @inbounds for i in 1:n
+        aii = A.data[i,i]
+        iszero(aii) && throw(SingularException(i))
+        bi = b[i]
+        for j in 1:i-1
+            bi -= A.data[i,j] * c[j]
         end
+        c[i] = aii \ bi
     end
-    return b
+    return c
 end
-function ldiv!(A::UnitLowerTriangular, b::AbstractVector)
-    require_one_based_indexing(A, b)
+function ldiv!(c::AbstractVector, A::UnitLowerTriangular, b::AbstractVector)
+    require_one_based_indexing(c, A, b)
     n = size(A, 2)
-    if !(n == length(b))
+    if n != length(b)
         throw(DimensionMismatch("second dimension of left hand side A, $n, and length of right hand side b, $(length(b)), must be equal"))
     end
-    @inbounds for j in 1:n
-        bj = b[j]
-        for i in j+1:n
-            b[i] -= A.data[i,j] * bj
+    if n != length(c)
+        throw(DimensionMismatch("length of output c, $(length(c)), does not match length of right hand side b, $(length(b))"))
+    end
+    @inbounds for i in 1:n
+        bi = b[i]
+        for j in 1:i-1
+            bi -= A.data[i,j] * c[j]
         end
+        c[i] = A[i,i] \ bi
```

Comment:
Isn't there a more efficient way to move the correct physical dimensions to `bi`?

When the element types of `A` and `b` are unitful integers, the result could also be unitful integers, without any floating point operations perfomed (no division by 1).

Reply:
I don't know of any. `convert` and alike fail because you can't convert, for instance, square meters to meters or a unitless number. Also, I'm not sure if there is a generic way of identifying whether a unitful quantity is integer-based. For unitless numbers you can see that I made an effort to preserve the integer return type.

Reply:
I worked around the issue. First, the target `c` is initialized with an integer unitful eltype, and second I replaced the division by a multiplication. I'll add a test for that also.

Reply:
Actually, no. For unitful quantities, we need to "perform" the division to get the right unit: assume both `A` and `b` have integer units of kilometers, then the result is "unit-free". In terms of types, this means `(Int*km) \ (Int*km) = Float64`, otherwise the kilometers don't cancel out.